[WebUI] Add save path, cookies and labels when adding torrents

naikel opened the pull request:
This patch allows people to configure save path, cookies and labels when adding torrents using the WebUI, either downloading from an URL or uploading it from the local filesystem.

This is a feature that is pretty popular in other bittorrents clients so I think qBittorrent being as great as it is should have them.

This consist in expanding the AddTorrentParams to allow passing cookies from the WebUI to the main program.

Also the webpages upload.html and download.html were changed a little bit to allow browsers to autocomplete.  The way they were designed doesn't allow browsers to autocomplete fields.  Nobody is going to paste cookies all the time in the WebUI they would want their browser to remember them.  This fix includes that as well.

glassez commented on src/core/bittorrent/session.h:
```diff
@@ -120,6 +121,7 @@ namespace BitTorrent
         QVector<int> filePriorities; // used if TorrentInfo is set
         bool ignoreShareRatio;
         bool skipChecking;
+        QList<QNetworkCookie> cookies;
```
Incorrect!
Cookies have nothing to add torrent to the session. Use DownloadManager::instance()->setCookiesFromUrl() directly.

glassez commented on src/core/bittorrent/session.cpp:
```diff
@@ -1002,7 +1003,8 @@ bool Session::addTorrent(QString source, const AddTorrentParams &params)
         Logger::instance()->addMessage(tr("Downloading '%1', please wait...", "e.g: Downloading 'xxx.torrent', please wait...").arg(source));
         // Launch downloader
-        Net::DownloadHandler *handler = Net::DownloadManager::instance()->downloadUrl(source, true, 10485760 /* 10MB */, true);
+        QList<QNetworkCookie> m_cookies = params.cookies;
```
Use DownloadManager::instance()->setCookiesFromUrl() directly. Don't touch (don't change) DownloadManager::downloadUrl() etc.

glassez commented on src/webui/prefjson.cpp:
```diff
@@ -141,7 +141,7 @@ QByteArray prefjson::getPreferences()
     data["max_ratio_enabled"] = (pref->getGlobalMaxRatio() >= 0.);
     data["max_ratio"] = pref->getGlobalMaxRatio();
-    data["max_ratio_act"] = static_cast<int>(pref->getMaxRatioAction());
+    data["max_ratio_act"] = QVariant::fromValue(pref->getMaxRatioAction());
```
You returned previously fixed bug (see #3939 and #3967).

glassez commented on src/webui/prefjson.cpp:
```diff
@@ -337,7 +337,7 @@ void prefjson::setPreferences(const QString& json)
         pref->setGlobalMaxRatio(-1);
     if (m.contains("max_ratio_act"))
-        pref->setMaxRatioAction(static_cast<MaxRatioAction>(m["max_ratio_act"].toInt()));
+        pref->setMaxRatioAction(m["max_ratio_act"].value<MaxRatioAction>());
```
Same as above.

glassez commented on src/webui/webapplication.h:
```diff
@@ -89,6 +89,7 @@ class WebApplication: public AbstractWebApplication
     void action_version_api();
     void action_version_api_min();
     void action_version_qbittorrent();
+    void action_command_getSavePath();
```
move this declaration to other action_command_* declarations.

naikel commented:
@glassez I fixed the implementation.  Actually I didn't know you could reference the DownloadManager from anywhere.  It looks way better now.

Also fixed the bugs returns.

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -318,7 +342,13 @@ void WebApplication::action_command_download()
+            params.savePath = savepath;
+            params.label = label;
+            const QUrl qurl = QUrl::fromEncoded(url.toUtf8());
+            Net::DownloadManager::instance()->setCookiesFromUrl(cookies,qurl);
```
space after comma

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -318,7 +342,13 @@ void WebApplication::action_command_download()
+            BitTorrent::AddTorrentParams params;
+            params.savePath = savepath;
+            params.label = label;
+            const QUrl qurl = QUrl::fromEncoded(url.toUtf8());
```
Don't create this const, use QUrl::fromEncoded(url.toUtf8() directly.

naikel replied:
I was tempted to do that but I checked the code style in downloadmanager.cpp in the function DownloadManager::downloadUrl where they did just that

```    
    // Process download request
    qDebug("url is %s", qPrintable(url));
    const QUrl qurl = QUrl::fromEncoded(url.toUtf8());
    QNetworkRequest request(qurl);
```

I'll fix it since I actually prefer not creating the qurl variable too.

glassez replied:
Existing code can also have problems...
And more. No need to create separate commits to fix your PR commits. Just rebase and correct them, otherwise it clogs up qBittorent history.

naikel replied:
Can you please elaborate a little more?  The only way to create a single commit in the Pull Request at this point would be to recreate the whole branch, that is, create it from master again, apply all the patches, a single commit, and then a push -f here.   Will that work?

naikel commented:
@glassez So I did what I think you meant.  git rebase -i, squashed all my commits in a single one, git push -f here.  Now there's a single commit.

glassez commented:
>So I did what I think you meant. git rebase -i, squashed all my commits in a single one, git push -f here. Now there's a single commit.

Basically, Yes.
But, for the future... It doesn't have to be a single commit. Just no need to split up your changes much. One commit for each set of logically separate changes. For example, in this PR it could be two commits: one for structural changes in the WebUI client code and another to implement support for cookies, save path, etc when adding a torrent.

naikel commented:
Noted.

So are you guys interested in this? my personal qBittorrent is heavily modded and I have many other features people would like to have.

sledgehammer999 commented:
>So are you guys interested in this?

Yes. But expect a little delay due to #3800.
The issues in the v3.3.0 milestone are "soft dependencies" except for the PRs. I'll start working on them this weekend and probably today too.

naikel commented:
I rebased this so it can be merged without conflict.

ngosang commented:
The Add torrent and Add link dialogs now looks ugly in the WebUI. Please, improve it.

naikel commented:
I'm not a designer so show me how you like them and I'll code it.  I just care for functionality not looks.

naikel commented:
I made a cosmetic fix following the style of the settings dialog.   Does it look better? 

![capture](https://cloud.githubusercontent.com/assets/7663686/11022698/3868f790-8637-11e5-8c92-3bf20b119e16.PNG)

ngosang commented:
Much better :smile: I have to test it carefully but the ui design is ok.

ngosang commented on src/webui/webapplication.cpp:
```diff
@@ -306,9 +308,31 @@ void WebApplication::action_command_shutdown()
 {
     CHECK_URI(0);
-    CHECK_PARAMETERS("urls");
+    CHECK_PARAMETERS("urls" << "savepath" << "label" << "cookie");
```
Whit this checks the 4 parameters are compulsory, so you are breaking backwards compatibility. This API is used by 3rd party apps. I prefer to make "urls" compulsory as before and the other parameters optional.
if not "savepath" set the default value
if not "label" set no label
if noy "cookie" set no cookie

ngosang replied:
I think you only need to remove the checks.
```
CHECK_PARAMETERS("urls");
```

naikel replied:
That won't work.  CHECK_PARAMETERS checks that exactly those parameters are in the request.  No more, no less.  I would have to delete the line completely, which is fine, and several other functions don't have it.

ngosang commented on src/webui/www/public/download.html:
```diff
@@ -5,15 +5,64 @@
-	</center>
+<fieldset class="settings" style="border: 0; text-align: left;">
+<div class="formRow" style="margin-top: 6px;">
+	<label for="savepath" class="leftLabelLarge">QBT_TR(Save files to location:)QBT_TR</label>
```
I think you need to add translatable strings `QBT_TR(XXXX)QBT_TR` to this file:
https://github.com/qbittorrent/qBittorrent/blob/master/src/webui/extra_translations.h

ngosang commented on src/webui/www/public/upload.html:
```diff
@@ -6,69 +6,61 @@
 </div>
-<div id="submitbutton" style="margin-top: 30px;">
+<div id="submitbutton" style="margin-top: 30px; text-align: center;">
 	<button type="submit" style="font-size: 1em;">QBT_TR(Upload Torrents)QBT_TR</button>
```
I know you didn't changed this but maybe we should change the button text to `QBT_TR(Download)QBT_TR`.

naikel replied:
I'm not sure about this change.  There are two options: Download from URL where qBittorrent downloads a torrent from a website, and Upload file, where the user uploads a file to qBittorrent.  So I think the name "Upload Torrents" is correct, but is my humble opinion though.

ngosang replied:
I know you didn't changed this but maybe we should change the button text to `QBT_TR(Download)QBT_TR`.

ngosang replied:
It's a minor thing, just wait for @sledgehammer999

sledgehammer999 replied:
And what is this?

naikel replied:
@sledgehammer999 that's the WebUI window to upload a torrent file.  The window title is "Download local torrent" but the button says "Upload Torrents", so it's weird.  I would change the title to Upload Torrents and @ngosang thinks the button should be changed to Download Torrents.

sledgehammer999 replied:
"Upload local torrent" conveys what is going on more accurately.

ngosang commented on src/webui/www/public/download.html:
```diff
@@ -5,15 +5,64 @@
+var submitted = false;
+
+$('downloadForm').addEventListener("submit", function() {
+        submitted = true;
```
Maybe we should add the spinner (loading image) as in the upload .torrent form. I tested it in this way.
1) Open the add link dialog.
2) Close qBittorret.
3) Send the form.

ngosang commented on src/webui/www/public/upload.html:
```diff
@@ -6,69 +6,61 @@
+        }).send();
+}
+
+getSavePath();
```
I'm not sure but I think it's safer to execute this call when the document is ready, the same for the other dialog.
https://learn.jquery.com/using-jquery-core/document-ready/

naikel replied:
I can do it like this:

```javascript
$(window).addEventListener("load", function() {
        getSavePath();
});
```

We can't use the jQuery's ready() function since we're not using that framework.  It has to be "pure javascript".

ngosang commented:
@naikel I let you some comments in the code. I tested everything except the cookies. If you implement my changes and you assure that the cookies work this PR can be merged.

naikel commented:
So I made the following changes:

1.- Changed the title of the upload window to "Upload local torrents"
2.- Added the spin to the download window
3.- Removed the checks of the parameters in the download window (the upload window didn't have the check in the first place).
4.- Included the new strings in extra_translations.h
5.- Changed the call of getSavePath() to a safer method where the whole window was loaded.

Cookies work, I've been using this patch since several years ago (since 3.1.9.2 actually).  It's just that recently I thought it should be in the main code.   I also have another good patch (seeding until x minutes have passed) that it's also useful to a lot of people that I will submit later.

lgallard commented:
Hi,

Will there be an API function to add torrents with savepath?

Regards,

naikel commented:
@lgallard With this patch you can send a command/download with savepath to whatever you want.  You can also set the label.

It also works with command/upload (savepath + label).

lgallard commented:
Excellent! People using the Android app have requested this!

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -306,9 +308,30 @@ void WebApplication::action_command_shutdown()
+
+        QStringList cookiesStr = cookie.split("; ");
+        foreach (QString cookieStr, cookiesStr)
+        {
```
Fix coding style.

naikel replied:
What do you mean exactly? Every file has a different coding style (weird I know), and the webapplication.cpp has
``` 
if () {
```
instead of
```
if () 
{
```
in all other methods.

Unless you mean something else...

glassez replied:
```c++
foreach (QString cookieString, cookiesString) {
}
```
See CODING_GUIDELINES.md

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -320,7 +343,12 @@ void WebApplication::action_command_download()
+            BitTorrent::AddTorrentParams params;
+            params.savePath = savepath;
+            params.label = label;
+            Net::DownloadManager::instance()->setCookiesFromUrl(cookies, QUrl::fromEncoded(url.toUtf8()));
```
Move this outside the loop.

glassez commented on src/webui/www/public/download.html:
```diff
@@ -4,16 +4,69 @@
+	        window.parent.closeWindows();
+});
+
+getSavePath = function() {
```
I prefer not to get involved in JavaScript code, but have you not accepted to follow the principle of code reuse?

glassez commented on src/webui/www/public/upload.html:
```diff
@@ -2,73 +2,67 @@
+		window.parent.closeWindows();
+});
+
+function getSavePath() {
```
This is the third repetition...

naikel commented:
More cosmetic fixes:

1.- Fixed the code style for the foreach.
2.- Moved getSavePath() to download.js

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -320,7 +346,9 @@ void WebApplication::action_command_download()
                 url = "magnet:?xt=urn:btih:" + url;
-            BitTorrent::Session::instance()->addTorrent(url);
+
+            Net::DownloadManager::instance()->setCookiesFromUrl(cookies, QUrl::fromEncoded(url.toUtf8()));
```
I say again: move this line outside the loop.

naikel replied:
You can't.  You need to set the cookie per URL and the url variable is inside the loop.  Remember you can download several URLs at the same time.

glassez replied:
Oh, I'm sorry. But then what's the point? You set the same cookies to different URLs. Bring an example of this. Why is it used?

naikel replied:
You can blame Qt for this.

The most common example is several torrent files from the same private tracker.  You'd think "well it's the same domain it should be the same cookie for all".  And you're right.

Sadly QNetworkCookie::setCookiesFromURL is "smart" and it asks for a whole URL instead of just the domain, and it adds the cookie to the cookieJar for the domain and "it did the work for you because you didn't need to parse the domain from the URL".  They actually say this with pride.

So basically if all the URLs are from the same domain, the first call will add it to the cookieJar and the next ones QNetworkCookie::setCookiesFromURL should notice it's already there for the same domain and just do nothing.

naikel commented:
Is there anything left to do for this PR or is it just waiting for a merge?

sledgehammer999 commented:
It is waiting review from me and merging. Since others have reviewed it should be done quickly. Just so you know, until the coming weekend I can't merge it. That is when v3.3.0 will be released and new features can flow again.

sledgehammer999 commented:
review done. Just wait for v3.3.0 and then this can get merged.

sledgehammer999 commented:
Thanks.

skyway22 commented:
so, is this implemented, or it will be? Cause in my webui there's nothing showing up..

thalieht commented:
It is. 3 3.4 is current version,

skyway22 commented:
yeah but linux arm was left behind.. 

> qBittorrent v3.1.10 (Web UI)
